File: cyber-cursor/backend/app/services/monitoring_siem_soar_service.py
import asyncio
import json
import uuid
from datetime import datetime, timedelta
from dataclasses import dataclass, field
from enum import Enum
from typing import List, Dict, Optional, Any
import structlog

logger = structlog.get_logger()

class LogLevel(Enum):
    DEBUG = "debug"
    INFO = "info"
    WARNING = "warning"
    ERROR = "error"
    CRITICAL = "critical"

class IncidentSeverity(Enum):
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"

class IncidentStatus(Enum):
    OPEN = "open"
    IN_PROGRESS = "in_progress"
    RESOLVED = "resolved"
    CLOSED = "closed"

class AlertType(Enum):
    SECURITY = "security"
    PERFORMANCE = "performance"
    AVAILABILITY = "availability"
    COMPLIANCE = "compliance"

class AnomalyType(Enum):
    BEHAVIORAL = "behavioral"
    NETWORK = "network"
    SYSTEM = "system"
    USER = "user"

class AutomationAction(Enum):
    BLOCK_IP = "block_ip"
    QUARANTINE_DEVICE = "quarantine_device"
    SEND_ALERT = "send_alert"
    CREATE_INCIDENT = "create_incident"
    ESCALATE = "escalate"

@dataclass
class LogEntry:
    id: str
    timestamp: datetime
    level: LogLevel
    source: str
    message: str
    metadata: Dict[str, Any] = field(default_factory=dict)
    user_id: Optional[str] = None
    session_id: Optional[str] = None
    ip_address: Optional[str] = None

@dataclass
class SecurityEvent:
    id: str
    timestamp: datetime
    event_type: str
    severity: IncidentSeverity
    source: str
    description: str
    raw_data: Dict[str, Any] = field(default_factory=dict)
    indicators: List[str] = field(default_factory=list)
    threat_intelligence: Dict[str, Any] = field(default_factory=dict)

@dataclass
class Incident:
    id: str
    title: str
    description: str
    severity: IncidentSeverity
    status: IncidentStatus
    created_at: datetime
    updated_at: datetime
    assigned_to: Optional[str] = None
    tags: List[str] = field(default_factory=list)
    related_events: List[str] = field(default_factory=list)
    playbook: Optional[str] = None
    sla_deadline: Optional[datetime] = None

@dataclass
class Alert:
    id: str
    title: str
    description: str
    alert_type: AlertType
    severity: IncidentSeverity
    timestamp: datetime
    source: str
    acknowledged: bool = False
    acknowledged_by: Optional[str] = None
    acknowledged_at: Optional[datetime] = None
    metadata: Dict[str, Any] = field(default_factory=dict)

@dataclass
class Anomaly:
    id: str
    anomaly_type: AnomalyType
    severity: IncidentSeverity
    timestamp: datetime
    source: str
    description: str
    confidence_score: float
    baseline_data: Dict[str, Any] = field(default_factory=dict)
    current_data: Dict[str, Any] = field(default_factory=dict)
    status: str = "detected"

@dataclass
class AutomationRule:
    id: str
    name: str
    description: str
    conditions: List[Dict[str, Any]]
    actions: List[AutomationAction]
    enabled: bool = True
    priority: int = 0
    created_at: datetime = field(default_factory=datetime.utcnow)

@dataclass
class Playbook:
    id: str
    name: str
    description: str
    steps: List[Dict[str, Any]]
    triggers: List[str]
    enabled: bool = True
    estimated_duration: int = 0  # minutes

@dataclass
class MonitoringSummary:
    total_logs: int
    total_events: int
    open_incidents: int
    active_alerts: int
    anomalies_detected: int
    automation_rules: int
    playbooks: int
    system_health: str
    last_updated: datetime

class MonitoringSIEMSOARService:
    def __init__(self):
        self.logs: List[LogEntry] = []
        self.security_events: List[SecurityEvent] = []
        self.incidents: List[Incident] = []
        self.alerts: List[Alert] = []
        self.anomalies: List[Anomaly] = []
        self.automation_rules: List[AutomationRule] = []
        self.playbooks: List[Playbook] = []
        self.background_tasks: List[asyncio.Task] = []
        self._running = False
        
        # Initialize sample data
        self._initialize_sample_data()
    
    def _initialize_sample_data(self):
        """Initialize sample data for demonstration"""
        # Sample logs
        for i in range(100):
            self.logs.append(LogEntry(
                id=str(uuid.uuid4()),
                timestamp=datetime.utcnow() - timedelta(minutes=i),
                level=LogLevel.INFO if i % 10 != 0 else LogLevel.ERROR,
                source=f"service-{i % 5}",
                message=f"Sample log message {i}",
                metadata={"request_id": f"req-{i}"},
                user_id=f"user-{i % 10}" if i % 3 == 0 else None
            ))
        
        # Sample security events
        event_types = ["failed_login", "suspicious_activity", "malware_detected", "data_exfiltration"]
        for i in range(20):
            self.security_events.append(SecurityEvent(
                id=str(uuid.uuid4()),
                timestamp=datetime.utcnow() - timedelta(hours=i),
                event_type=event_types[i % len(event_types)],
                severity=IncidentSeverity.HIGH if i % 5 == 0 else IncidentSeverity.MEDIUM,
                source=f"endpoint-{i % 10}",
                description=f"Security event {i} detected",
                indicators=[f"indicator-{i}"],
                threat_intelligence={"threat_score": 75 + i}
            ))
        
        # Sample incidents
        for i in range(5):
            self.incidents.append(Incident(
                id=str(uuid.uuid4()),
                title=f"Security Incident {i + 1}",
                description=f"Description for incident {i + 1}",
                severity=IncidentSeverity.HIGH if i % 2 == 0 else IncidentSeverity.MEDIUM,
                status=IncidentStatus.OPEN if i < 3 else IncidentStatus.IN_PROGRESS,
                created_at=datetime.utcnow() - timedelta(hours=i * 2),
                updated_at=datetime.utcnow() - timedelta(hours=i),
                assigned_to=f"analyst-{i % 3}" if i > 1 else None,
                tags=["malware", "endpoint"] if i % 2 == 0 else ["network", "suspicious"]
            ))
        
        # Sample alerts
        for i in range(15):
            self.alerts.append(Alert(
                id=str(uuid.uuid4()),
                title=f"Alert {i + 1}",
                description=f"Alert description {i + 1}",
                alert_type=AlertType.SECURITY if i % 3 == 0 else AlertType.PERFORMANCE,
                severity=IncidentSeverity.MEDIUM,
                timestamp=datetime.utcnow() - timedelta(minutes=i * 30),
                source=f"monitor-{i % 5}",
                acknowledged=i > 10
            ))
        
        # Sample anomalies
        for i in range(8):
            self.anomalies.append(Anomaly(
                id=str(uuid.uuid4()),
                anomaly_type=AnomalyType.BEHAVIORAL if i % 2 == 0 else AnomalyType.NETWORK,
                severity=IncidentSeverity.MEDIUM,
                timestamp=datetime.utcnow() - timedelta(hours=i),
                source=f"user-{i % 10}",
                description=f"Anomaly detected: {i + 1}",
                confidence_score=0.85 + (i * 0.02)
            ))
        
        # Sample automation rules
        self.automation_rules.extend([
            AutomationRule(
                id=str(uuid.uuid4()),
                name="Block Suspicious IP",
                description="Automatically block IPs with high threat score",
                conditions=[{"field": "threat_score", "operator": ">", "value": 80}],
                actions=[AutomationAction.BLOCK_IP, AutomationAction.CREATE_INCIDENT],
                priority=1
            ),
            AutomationRule(
                id=str(uuid.uuid4()),
                name="Quarantine Malware",
                description="Quarantine devices with detected malware",
                conditions=[{"field": "event_type", "operator": "==", "value": "malware_detected"}],
                actions=[AutomationAction.QUARANTINE_DEVICE, AutomationAction.SEND_ALERT],
                priority=2
            )
        ])
        
        # Sample playbooks
        self.playbooks.extend([
            Playbook(
                id=str(uuid.uuid4()),
                name="Malware Response",
                description="Standard response for malware incidents",
                steps=[
                    {"step": 1, "action": "Isolate affected system", "duration": 5},
                    {"step": 2, "action": "Collect evidence", "duration": 15},
                    {"step": 3, "action": "Remove malware", "duration": 30},
                    {"step": 4, "action": "Verify system integrity", "duration": 10}
                ],
                triggers=["malware_detected"],
                estimated_duration=60
            ),
            Playbook(
                id=str(uuid.uuid4()),
                name="Data Breach Response",
                description="Response procedure for data breach incidents",
                steps=[
                    {"step": 1, "action": "Assess scope of breach", "duration": 30},
                    {"step": 2, "action": "Notify stakeholders", "duration": 15},
                    {"step": 3, "action": "Implement containment", "duration": 45},
                    {"step": 4, "action": "Document incident", "duration": 20}
                ],
                triggers=["data_exfiltration"],
                estimated_duration=110
            )
        ])
    
    async def start_monitoring_siem_soar_service(self):
        """Start the monitoring, SIEM & SOAR service"""
        if self._running:
            return
        
        self._running = True
        logger.info("Starting Monitoring, SIEM & SOAR service")
        
        # Start background tasks
        self.background_tasks.extend([
            asyncio.create_task(self._log_processor()),
            asyncio.create_task(self._event_correlator()),
            asyncio.create_task(self._anomaly_detector()),
            asyncio.create_task(self._automation_engine())
        ])
        
        logger.info("Monitoring, SIEM & SOAR service started successfully")
    
    async def stop_monitoring_siem_soar_service(self):
        """Stop the monitoring, SIEM & SOAR service"""
        if not self._running:
            return
        
        self._running = False
        logger.info("Stopping Monitoring, SIEM & SOAR service")
        
        # Cancel background tasks
        for task in self.background_tasks:
            task.cancel()
        
        await asyncio.gather(*self.background_tasks, return_exceptions=True)
        self.background_tasks.clear()
        
        logger.info("Monitoring, SIEM & SOAR service stopped")
    
    async def _log_processor(self):
        """Background task for processing logs"""
        while self._running:
            try:
                # Simulate log processing
                await asyncio.sleep(5)
                logger.debug("Processing logs...")
            except asyncio.CancelledError:
                break
            except Exception as e:
                logger.error(f"Error in log processor: {e}")
    
    async def _event_correlator(self):
        """Background task for correlating security events"""
        while self._running:
            try:
                # Simulate event correlation
                await asyncio.sleep(10)
                logger.debug("Correlating security events...")
            except asyncio.CancelledError:
                break
            except Exception as e:
                logger.error(f"Error in event correlator: {e}")
    
    async def _anomaly_detector(self):
        """Background task for detecting anomalies"""
        while self._running:
            try:
                # Simulate anomaly detection
                await asyncio.sleep(15)
                logger.debug("Detecting anomalies...")
            except asyncio.CancelledError:
                break
            except Exception as e:
                logger.error(f"Error in anomaly detector: {e}")
    
    async def _automation_engine(self):
        """Background task for executing automation rules"""
        while self._running:
            try:
                # Simulate automation execution
                await asyncio.sleep(20)
                logger.debug("Executing automation rules...")
            except asyncio.CancelledError:
                break
            except Exception as e:
                logger.error(f"Error in automation engine: {e}")
    
    # Log Management
    async def add_log(self, level: LogLevel, source: str, message: str, 
                     metadata: Dict[str, Any] = None, user_id: str = None,
                     session_id: str = None, ip_address: str = None) -> LogEntry:
        """Add a new log entry"""
        log_entry = LogEntry(
            id=str(uuid.uuid4()),
            timestamp=datetime.utcnow(),
            level=level,
            source=source,
            message=message,
            metadata=metadata or {},
            user_id=user_id,
            session_id=session_id,
            ip_address=ip_address
        )
        
        self.logs.append(log_entry)
        logger.info(f"Added log entry: {log_entry.id}")
        return log_entry
    
    async def get_logs(self, level: LogLevel = None, source: str = None,
                      start_time: datetime = None, end_time: datetime = None,
                      limit: int = 100) -> List[LogEntry]:
        """Get logs with optional filtering"""
        filtered_logs = self.logs
        
        if level:
            filtered_logs = [log for log in filtered_logs if log.level == level]
        if source:
            filtered_logs = [log for log in filtered_logs if log.source == source]
        if start_time:
            filtered_logs = [log for log in filtered_logs if log.timestamp >= start_time]
        if end_time:
            filtered_logs = [log for log in filtered_logs if log.timestamp <= end_time]
        
        return filtered_logs[-limit:]
    
    # Security Event Management
    async def add_security_event(self, event_type: str, severity: IncidentSeverity,
                                source: str, description: str, raw_data: Dict[str, Any] = None,
                                indicators: List[str] = None, threat_intelligence: Dict[str, Any] = None) -> SecurityEvent:
        """Add a new security event"""
        event = SecurityEvent(
            id=str(uuid.uuid4()),
            timestamp=datetime.utcnow(),
            event_type=event_type,
            severity=severity,
            source=source,
            description=description,
            raw_data=raw_data or {},
            indicators=indicators or [],
            threat_intelligence=threat_intelligence or {}
        )
        
        self.security_events.append(event)
        logger.info(f"Added security event: {event.id}")
        return event
    
    async def get_security_events(self, event_type: str = None, severity: IncidentSeverity = None,
                                 start_time: datetime = None, end_time: datetime = None,
                                 limit: int = 100) -> List[SecurityEvent]:
        """Get security events with optional filtering"""
        filtered_events = self.security_events
        
        if event_type:
            filtered_events = [event for event in filtered_events if event.event_type == event_type]
        if severity:
            filtered_events = [event for event in filtered_events if event.severity == severity]
        if start_time:
            filtered_events = [event for event in filtered_events if event.timestamp >= start_time]
        if end_time:
            filtered_events = [event for event in filtered_events if event.timestamp <= end_time]
        
        return filtered_events[-limit:]
    
    # Incident Management
    async def create_incident(self, title: str, description: str, severity: IncidentSeverity,
                             tags: List[str] = None, playbook: str = None) -> Incident:
        """Create a new incident"""
        incident = Incident(
            id=str(uuid.uuid4()),
            title=title,
            description=description,
            severity=severity,
            status=IncidentStatus.OPEN,
            created_at=datetime.utcnow(),
            updated_at=datetime.utcnow(),
            tags=tags or [],
            playbook=playbook
        )
        
        self.incidents.append(incident)
        logger.info(f"Created incident: {incident.id}")
        return incident
    
    async def update_incident(self, incident_id: str, status: IncidentStatus = None,
                             assigned_to: str = None, tags: List[str] = None) -> Optional[Incident]:
        """Update an incident"""
        for incident in self.incidents:
            if incident.id == incident_id:
                if status:
                    incident.status = status
                if assigned_to:
                    incident.assigned_to = assigned_to
                if tags:
                    incident.tags = tags
                
                incident.updated_at = datetime.utcnow()
                logger.info(f"Updated incident: {incident_id}")
                return incident
        
        return None
    
    async def get_incidents(self, status: IncidentStatus = None, severity: IncidentSeverity = None,
                           assigned_to: str = None, limit: int = 100) -> List[Incident]:
        """Get incidents with optional filtering"""
        filtered_incidents = self.incidents
        
        if status:
            filtered_incidents = [incident for incident in filtered_incidents if incident.status == status]
        if severity:
            filtered_incidents = [incident for incident in filtered_incidents if incident.severity == severity]
        if assigned_to:
            filtered_incidents = [incident for incident in filtered_incidents if incident.assigned_to == assigned_to]
        
        return filtered_incidents[-limit:]
    
    # Alert Management
    async def create_alert(self, title: str, description: str, alert_type: AlertType,
                          severity: IncidentSeverity, source: str, metadata: Dict[str, Any] = None) -> Alert:
        """Create a new alert"""
        alert = Alert(
            id=str(uuid.uuid4()),
            title=title,
            description=description,
            alert_type=alert_type,
            severity=severity,
            timestamp=datetime.utcnow(),
            source=source,
            metadata=metadata or {}
        )
        
        self.alerts.append(alert)
        logger.info(f"Created alert: {alert.id}")
        return alert
    
    async def acknowledge_alert(self, alert_id: str, acknowledged_by: str) -> Optional[Alert]:
        """Acknowledge an alert"""
        for alert in self.alerts:
            if alert.id == alert_id:
                alert.acknowledged = True
                alert.acknowledged_by = acknowledged_by
                alert.acknowledged_at = datetime.utcnow()
                logger.info(f"Acknowledged alert: {alert_id}")
                return alert
        
        return None
    
    async def get_alerts(self, alert_type: AlertType = None, severity: IncidentSeverity = None,
                        acknowledged: bool = None, limit: int = 100) -> List[Alert]:
        """Get alerts with optional filtering"""
        filtered_alerts = self.alerts
        
        if alert_type:
            filtered_alerts = [alert for alert in filtered_alerts if alert.alert_type == alert_type]
        if severity:
            filtered_alerts = [alert for alert in filtered_alerts if alert.severity == severity]
        if acknowledged is not None:
            filtered_alerts = [alert for alert in filtered_alerts if alert.acknowledged == acknowledged]
        
        return filtered_alerts[-limit:]
    
    # Anomaly Management
    async def add_anomaly(self, anomaly_type: AnomalyType, severity: IncidentSeverity,
                         source: str, description: str, confidence_score: float,
                         baseline_data: Dict[str, Any] = None, current_data: Dict[str, Any] = None) -> Anomaly:
        """Add a new anomaly"""
        anomaly = Anomaly(
            id=str(uuid.uuid4()),
            anomaly_type=anomaly_type,
            severity=severity,
            timestamp=datetime.utcnow(),
            source=source,
            description=description,
            confidence_score=confidence_score,
            baseline_data=baseline_data or {},
            current_data=current_data or {}
        )
        
        self.anomalies.append(anomaly)
        logger.info(f"Added anomaly: {anomaly.id}")
        return anomaly
    
    async def get_anomalies(self, anomaly_type: AnomalyType = None, severity: IncidentSeverity = None,
                           start_time: datetime = None, end_time: datetime = None,
                           limit: int = 100) -> List[Anomaly]:
        """Get anomalies with optional filtering"""
        filtered_anomalies = self.anomalies
        
        if anomaly_type:
            filtered_anomalies = [anomaly for anomaly in filtered_anomalies if anomaly.anomaly_type == anomaly_type]
        if severity:
            filtered_anomalies = [anomaly for anomaly in filtered_anomalies if anomaly.severity == severity]
        if start_time:
            filtered_anomalies = [anomaly for anomaly in filtered_anomalies if anomaly.timestamp >= start_time]
        if end_time:
            filtered_anomalies = [anomaly for anomaly in filtered_anomalies if anomaly.timestamp <= end_time]
        
        return filtered_anomalies[-limit:]
    
    # Automation Rule Management
    async def create_automation_rule(self, name: str, description: str, conditions: List[Dict[str, Any]],
                                   actions: List[AutomationAction], priority: int = 0) -> AutomationRule:
        """Create a new automation rule"""
        rule = AutomationRule(
            id=str(uuid.uuid4()),
            name=name,
            description=description,
            conditions=conditions,
            actions=actions,
            priority=priority
        )
        
        self.automation_rules.append(rule)
        logger.info(f"Created automation rule: {rule.id}")
        return rule
    
    async def get_automation_rules(self, enabled: bool = None) -> List[AutomationRule]:
        """Get automation rules with optional filtering"""
        filtered_rules = self.automation_rules
        
        if enabled is not None:
            filtered_rules = [rule for rule in filtered_rules if rule.enabled == enabled]
        
        return sorted(filtered_rules, key=lambda x: x.priority, reverse=True)
    
    # Playbook Management
    async def create_playbook(self, name: str, description: str, steps: List[Dict[str, Any]],
                             triggers: List[str], estimated_duration: int = 0) -> Playbook:
        """Create a new playbook"""
        playbook = Playbook(
            id=str(uuid.uuid4()),
            name=name,
            description=description,
            steps=steps,
            triggers=triggers,
            estimated_duration=estimated_duration
        )
        
        self.playbooks.append(playbook)
        logger.info(f"Created playbook: {playbook.id}")
        return playbook
    
    async def get_playbooks(self, enabled: bool = None) -> List[Playbook]:
        """Get playbooks with optional filtering"""
        filtered_playbooks = self.playbooks
        
        if enabled is not None:
            filtered_playbooks = [playbook for playbook in filtered_playbooks if playbook.enabled == enabled]
        
        return filtered_playbooks
    
    # Summary and Analytics
    async def get_monitoring_summary(self) -> MonitoringSummary:
        """Get monitoring summary statistics"""
        open_incidents = len([incident for incident in self.incidents if incident.status == IncidentStatus.OPEN])
        active_alerts = len([alert for alert in self.alerts if not alert.acknowledged])
        recent_anomalies = len([anomaly for anomaly in self.anomalies 
                               if anomaly.timestamp > datetime.utcnow() - timedelta(hours=24)])
        
        # Determine system health based on various metrics
        system_health = "healthy"
        if open_incidents > 5 or active_alerts > 10:
            system_health = "warning"
        if open_incidents > 10 or active_alerts > 20:
            system_health = "critical"
        
        return MonitoringSummary(
            total_logs=len(self.logs),
            total_events=len(self.security_events),
            open_incidents=open_incidents,
            active_alerts=active_alerts,
            anomalies_detected=recent_anomalies,
            automation_rules=len([rule for rule in self.automation_rules if rule.enabled]),
            playbooks=len([playbook for playbook in self.playbooks if playbook.enabled]),
            system_health=system_health,
            last_updated=datetime.utcnow()
        )

# Global service instance
monitoring_siem_soar_service = MonitoringSIEMSOARService() 

File: cyber-cursor/frontend/src/services/monitoringSiemSoarService.ts
import { apiClient } from '../utils/apiClient';

export interface LogEntry {
  id: string;
  timestamp: string;
  source: string;
  level: 'debug' | 'info' | 'warning' | 'error' | 'critical';
  message: string;
  details: Record<string, any>;
  correlation_id?: string;
  user_id?: string;
  ip_address?: string;
  session_id?: string;
}

export interface SecurityEvent {
  id: string;
  timestamp: string;
  event_type: string;
  severity: 'low' | 'medium' | 'high' | 'critical';
  source: string;
  description: string;
  affected_assets: string[];
  indicators: string[];
  status: 'new' | 'investigating' | 'resolved' | 'false_positive';
  assigned_to?: string;
  tags: string[];
  metadata: Record<string, any>;
}

export interface Alert {
  id: string;
  timestamp: string;
  alert_type: string;
  severity: 'low' | 'medium' | 'high' | 'critical';
  source: string;
  title: string;
  description: string;
  affected_assets: string[];
  status: 'active' | 'acknowledged' | 'resolved' | 'suppressed';
  assigned_to?: string;
  escalation_level: number;
  auto_close: boolean;
  metadata: Record<string, any>;
}

export interface Incident {
  id: string;
  incident_number: string;
  title: string;
  description: string;
  severity: 'low' | 'medium' | 'high' | 'critical';
  status: 'open' | 'investigating' | 'contained' | 'resolved' | 'closed';
  priority: 'low' | 'medium' | 'high' | 'urgent';
  category: string;
  subcategory: string;
  affected_assets: string[];
  assigned_team: string;
  assigned_to?: string;
  created_at: string;
  updated_at: string;
  resolved_at?: string;
  sla_target?: string;
  sla_breach: boolean;
  tags: string[];
  playbook_id?: string;
  timeline: IncidentTimelineEntry[];
}

export interface IncidentTimelineEntry {
  id: string;
  timestamp: string;
  action: string;
  description: string;
  user: string;
  details?: Record<string, any>;
}

export interface Playbook {
  id: string;
  name: string;
  description: string;
  category: string;
  severity_levels: string[];
  steps: PlaybookStep[];
  estimated_duration: string;
  required_skills: string[];
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface PlaybookStep {
  id: string;
  order: number;
  title: string;
  description: string;
  action_type: string;
  parameters: Record<string, any>;
  expected_outcome: string;
  timeout_minutes: number;
  is_manual: boolean;
  dependencies: string[];
}

export interface AutomationRule {
  id: string;
  name: string;
  description: string;
  trigger_conditions: TriggerCondition[];
  actions: AutomationAction[];
  is_active: boolean;
  priority: number;
  created_at: string;
  updated_at: string;
}

export interface TriggerCondition {
  field: string;
  operator: 'equals' | 'contains' | 'regex' | 'greater_than' | 'less_than';
  value: any;
  logical_operator?: 'AND' | 'OR';
}

export interface AutomationAction {
  id: string;
  action_type: string;
  parameters: Record<string, any>;
  order: number;
  timeout_seconds?: number;
  retry_count?: number;
}

export interface ThreatIntelligence {
  id: string;
  indicator: string;
  indicator_type: string;
  threat_type: string;
  confidence: number;
  severity: 'low' | 'medium' | 'high' | 'critical';
  source: string;
  first_seen: string;
  last_seen: string;
  tags: string[];
  metadata: Record<string, any>;
}

export interface Dashboard {
  id: string;
  name: string;
  description: string;
  widgets: DashboardWidget[];
  layout: any;
  is_default: boolean;
  created_at: string;
  updated_at: string;
}

export interface DashboardWidget {
  id: string;
  type: string;
  title: string;
  config: Record<string, any>;
  position: { x: number; y: number; w: number; h: number };
}

export interface MetricsData {
  metric_name: string;
  values: Array<{ timestamp: string; value: number }>;
  unit: string;
  trend: 'up' | 'down' | 'stable';
}

class MonitoringSiemSoarService {
  private baseUrl = '/api/v1/monitoring-siem-soar';

  // Log Management
  async getLogs(
    startTime?: string,
    endTime?: string,
    level?: string,
    source?: string,
    limit: number = 100
  ): Promise<LogEntry[]> {
    const params = new URLSearchParams();
    if (startTime) params.append('start_time', startTime);
    if (endTime) params.append('end_time', endTime);
    if (level) params.append('level', level);
    if (source) params.append('source', source);
    params.append('limit', limit.toString());
    
    const response = await apiClient.get(`${this.baseUrl}/logs?${params.toString()}`);
    return response.data;
  }

  async searchLogs(query: string, filters?: Record<string, any>): Promise<LogEntry[]> {
    const response = await apiClient.post(`${this.baseUrl}/logs/search`, {
      query,
      filters
    });
    return response.data;
  }

  // Security Events
  async getSecurityEvents(
    severity?: string,
    status?: string,
    eventType?: string,
    limit: number = 100
  ): Promise<SecurityEvent[]> {
    const params = new URLSearchParams();
    if (severity) params.append('severity', severity);
    if (status) params.append('status', status);
    if (eventType) params.append('event_type', eventType);
    params.append('limit', limit.toString());
    
    const response = await apiClient.get(`${this.baseUrl}/security-events?${params.toString()}`);
    return response.data;
  }

  async updateSecurityEventStatus(eventId: string, status: string, notes?: string): Promise<{
    message: string;
    event_id: string;
    status: string;
  }> {
    const response = await apiClient.patch(`${this.baseUrl}/security-events/${eventId}`, {
      status,
      notes
    });
    return response.data;
  }

  // Alert Management
  async getAlerts(
    severity?: string,
    status?: string,
    alertType?: string,
    limit: number = 100
  ): Promise<Alert[]> {
    const params = new URLSearchParams();
    if (severity) params.append('severity', severity);
    if (status) params.append('status', status);
    if (alertType) params.append('alert_type', alertType);
    params.append('limit', limit.toString());
    
    const response = await apiClient.get(`${this.baseUrl}/alerts?${params.toString()}`);
    return response.data;
  }

  async acknowledgeAlert(alertId: string, userId: string, notes?: string): Promise<{
    message: string;
    alert_id: string;
    status: string;
  }> {
    const response = await apiClient.post(`${this.baseUrl}/alerts/${alertId}/acknowledge`, {
      user_id: userId,
      notes
    });
    return response.data;
  }

  async resolveAlert(alertId: string, userId: string, resolution: string): Promise<{
    message: string;
    alert_id: string;
    status: string;
  }> {
    const response = await apiClient.post(`${this.baseUrl}/alerts/${alertId}/resolve`, {
      user_id: userId,
      resolution
    });
    return response.data;
  }

  // Incident Management
  async getIncidents(
    status?: string,
    severity?: string,
    priority?: string,
    limit: number = 100
  ): Promise<Incident[]> {
    const params = new URLSearchParams();
    if (status) params.append('status', status);
    if (severity) params.append('severity', severity);
    if (priority) params.append('priority', priority);
    params.append('limit', limit.toString());
    
    const response = await apiClient.get(`${this.baseUrl}/incidents?${params.toString()}`);
    return response.data;
  }

  async createIncident(incidentData: Partial<Incident>): Promise<{
    message: string;
    incident_id: string;
    incident_number: string;
  }> {
    const response = await apiClient.post(`${this.baseUrl}/incidents`, incidentData);
    return response.data;
  }

  async updateIncident(incidentId: string, updateData: Partial<Incident>): Promise<{
    message: string;
    incident_id: string;
    status: string;
  }> {
    const response = await apiClient.patch(`${this.baseUrl}/incidents/${incidentId}`, updateData);
    return response.data;
  }

  async addIncidentTimelineEntry(
    incidentId: string,
    entry: Omit<IncidentTimelineEntry, 'id'>
  ): Promise<{
    message: string;
    entry_id: string;
  }> {
    const response = await apiClient.post(`${this.baseUrl}/incidents/${incidentId}/timeline`, entry);
    return response.data;
  }

  // Playbook Management
  async getPlaybooks(category?: string, isActive?: boolean): Promise<Playbook[]> {
    const params = new URLSearchParams();
    if (category) params.append('category', category);
    if (isActive !== undefined) params.append('is_active', isActive.toString());
    
    const response = await apiClient.get(`${this.baseUrl}/playbooks?${params.toString()}`);
    return response.data;
  }

  async executePlaybook(
    playbookId: string,
    incidentId: string,
    parameters?: Record<string, any>
  ): Promise<{
    message: string;
    execution_id: string;
    status: string;
    estimated_duration: string;
  }> {
    const response = await apiClient.post(`${this.baseUrl}/playbooks/${playbookId}/execute`, {
      incident_id: incidentId,
      parameters
    });
    return response.data;
  }

  // Automation Rules
  async getAutomationRules(isActive?: boolean): Promise<AutomationRule[]> {
    const params = new URLSearchParams();
    if (isActive !== undefined) params.append('is_active', isActive.toString());
    
    const response = await apiClient.get(`${this.baseUrl}/automation-rules?${params.toString()}`);
    return response.data;
  }

  async createAutomationRule(ruleData: Omit<AutomationRule, 'id' | 'created_at' | 'updated_at'>): Promise<{
    message: string;
    rule_id: string;
  }> {
    const response = await apiClient.post(`${this.baseUrl}/automation-rules`, ruleData);
    return response.data;
  }

  async updateAutomationRule(ruleId: string, updateData: Partial<AutomationRule>): Promise<{
    message: string;
    rule_id: string;
    status: string;
  }> {
    const response = await apiClient.patch(`${this.baseUrl}/automation-rules/${ruleId}`, updateData);
    return response.data;
  }

  // Threat Intelligence
  async getThreatIntelligence(
    indicatorType?: string,
    threatType?: string,
    confidence?: number,
    limit: number = 100
  ): Promise<ThreatIntelligence[]> {
    const params = new URLSearchParams();
    if (indicatorType) params.append('indicator_type', indicatorType);
    if (threatType) params.append('threat_type', threatType);
    if (confidence) params.append('confidence', confidence.toString());
    params.append('limit', limit.toString());
    
    const response = await apiClient.get(`${this.baseUrl}/threat-intelligence?${params.toString()}`);
    return response.data;
  }

  async addThreatIntelligence(intelData: Omit<ThreatIntelligence, 'id'>): Promise<{
    message: string;
    intel_id: string;
  }> {
    const response = await apiClient.post(`${this.baseUrl}/threat-intelligence`, intelData);
    return response.data;
  }

  // Dashboards
  async getDashboards(): Promise<Dashboard[]> {
    const response = await apiClient.get(`${this.baseUrl}/dashboards`);
    return response.data;
  }

  async getDashboardData(dashboardId: string): Promise<{
    dashboard: Dashboard;
    data: Record<string, any>;
  }> {
    const response = await apiClient.get(`${this.baseUrl}/dashboards/${dashboardId}/data`);
    return response.data;
  }

  // Metrics and Analytics
  async getMetrics(
    metricNames: string[],
    startTime: string,
    endTime: string,
    interval: string = '1h'
  ): Promise<MetricsData[]> {
    const response = await apiClient.post(`${this.baseUrl}/metrics`, {
      metric_names: metricNames,
      start_time: startTime,
      end_time: endTime,
      interval
    });
    return response.data;
  }

  async getRealTimeMetrics(metricNames: string[]): Promise<Record<string, number>> {
    const response = await apiClient.post(`${this.baseUrl}/metrics/realtime`, {
      metric_names: metricNames
    });
    return response.data;
  }

  // Correlation and Analysis
  async correlateEvents(eventIds: string[]): Promise<{
    correlation_id: string;
    events: SecurityEvent[];
    patterns: string[];
    risk_score: number;
    recommendations: string[];
  }> {
    const response = await apiClient.post(`${this.baseUrl}/correlation`, {
      event_ids: eventIds
    });
    return response.data;
  }

  async getEventPatterns(
    startTime: string,
    endTime: string,
    patternType?: string
  ): Promise<{
    patterns: Array<{
      pattern: string;
      frequency: number;
      severity: string;
      affected_assets: string[];
      first_seen: string;
      last_seen: string;
    }>;
  }> {
    const params = new URLSearchParams();
    params.append('start_time', startTime);
    params.append('end_time', endTime);
    if (patternType) params.append('pattern_type', patternType);
    
    const response = await apiClient.get(`${this.baseUrl}/patterns?${params.toString()}`);
    return response.data;
  }
}

export const monitoringSiemSoarService = new MonitoringSiemSoarService();
